functorch/experimental/_map.py

```diff
@@ -45,69 +100,126 @@ def trace_map(proxy_mode, func_overload, f, xs, *args):
             next_name = candidate
 
     proxy_mode.tracer.root.register_module(next_name, body_graph)
-    node_args = (body_graph, xs, *args)
+    node_args = (body_graph, num_mapped, *args)
     proxy_args = pytree.tree_map(partial(unwrap_proxy, proxy_mode), node_args)
     out_proxy = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, {},
-                                               name="map")
-    outs = [body_graph(x, *args) for x in xs]
-    # Implementation notes: we need to use new_empty() + copy_() here instead of stack() directly
-    # because stack([...]) takes a fixed size list which will specialize dynamic shape here.
-    # Meanwhile we want to preserve the looped over dimension as symbolic shape, such that:
-    # ys: Tensor[s0, ...] = map(xs: Tensor[s0, ...], *args)
-    out = outs[0].new_empty([xs.shape[0], *outs[0].shape])
-    out.copy_(torch.stack(outs))
-    return track_tensor_tree(out, out_proxy, constant=None, tracer=proxy_mode.tracer)
-
-
-@map.py_impl(DispatchKey.CompositeExplicitAutograd)
-def map_cpu(f, xs, *args):
+                                               name="map_impl")
+    return track_tensor_tree(expanded_outs, out_proxy, constant=None, tracer=proxy_mode.tracer)
+
+def _unstack_pytree(xs):
+    flat_xs, inspec = pytree.tree_flatten(xs)
+    assert all(isinstance(xs, torch.Tensor) for xs in flat_xs), f"Leaves of xs must be Tensor {flat_xs}"
+    assert all(xs.shape[0] == flat_xs[0].shape[0] for xs in flat_xs), \
+        f"Leaves of xs must have same leading dimension size {[xs.shape for xs in flat_xs]}"
+    a = list(zip(*flat_xs))
+    pytrees = []
+    for tuple in a:
+        pytrees.append(pytree.tree_unflatten(tuple, inspec))
+    return pytrees
+
+def _stack_pytree(pytrees):
+    flat_out = []
+    out_spec = None
+    for pt in pytrees:
+        flat_pt, out_spec = pytree.tree_flatten(pt)
+        flat_out.append(flat_pt)
+    b = list(zip(*flat_out))
+    stacked_out = []
+    for leaves in b:
+        if all(leave is not None for leave in leaves):
+            stacked_out.append(torch.stack(leaves))
+        else:
+            stacked_out.append(None)
```

Comment:
Er, why can the output be None?

Reply:
i think it's due to some weried case I encounterd but I cannot recall now.. Maybe we should check the input and just do the stack here.

Reply:
We should probably just delete this branch and then put it back if/when we actually encounter the edge case. Reading your code, this should never happen, because the output of the body function passed to map must always be Tensors